src/packet/secret_key_macro.rs

```diff
@@ -151,30 +151,44 @@ macro_rules! impl_secret_key {
                     .sign_key(key, key_pw, &self)
             }
 
-            pub fn set_password<O, N, R>(
+            /// Lock this Secret Key packet with the password `pw_new`.
+            /// If `pw_new` is `None`, the private key material is unlocked.
+            ///
+            /// If the Secret Key packet is locked, the current password must be provided with
+            /// `pw_old`.
+            ///
+            /// This function uses the default s2k mechanism based on the key version
+            /// (`Cfb` with iterated and salted derivation of the password).
+            pub fn set_password<P, R>(
                 &mut self,
-                pw_old: O,
-                pw_new: N,
+                pw_old: P,
+                pw_new: Option<P>,
                 rng: R,
             ) -> $crate::errors::Result<()>
             where
-                O: FnOnce() -> String,
-                N: FnOnce() -> Option<String>,
+                P: FnOnce() -> String,
                 R: rand::Rng + rand::CryptoRng,
             {
                 let s2k = $crate::types::S2kParams::new_default(rng);
                 Self::set_password_and_s2k(self, pw_old, pw_new, s2k)
             }
 
-            pub fn set_password_and_s2k<O, N>(
+            /// Lock this Secret Key packet with the password `pw_new`.
+            /// If `pw_new` is `None`, the private key material is unlocked.
+            ///
+            /// If the Secret Key packet is locked, the current password must be provided with
+            /// `pw_old`.
+            ///
+            /// The s2k mechanism is set to `s2k_params`, if `pw_new` is `Some`.
+            /// If `pw_new` is `None`, `s2k_params` is not used.
+            pub fn set_password_and_s2k<P>(
                 &mut self,
-                pw_old: O,
-                pw_new: N,
+                pw_old: P,
+                pw_new: Option<P>,
```

Comment:
I am tempted to change the API to

`pw_new: Option<(P, $crate::types::S2kParams)>`

Tuples are a bit odd as fn parameters. But it would remove the weird situation of passing an `S2kParams` that is not used when `pw_new` is `None`, at the type level.

wdyt?

Reply:
I wonder if it isn't easier at this point to have a specific call for the `None` case

Reply:
Good point. I've pushed a change that explores that approach.

I'm half tempted to drop `pw_old` from the setter functions, now.
Callers could then unlock and lock explicitly, as needed.

Reply:
Pushed a commit that completely separates locking and unlocking.
I like that change - in part because it allows aligning function names with the terms in the RFC.